Adding FT3 detector for ALICE3 EndCaps simulation

rpezzi opened the pull request:
This PR creates simple forward layer detector named FT3 to get started with ALICE3 endcaps simulation. This is a follow up to PR #3910. FT3 is based on the ITSMFT classes, much like the ALICE3 TRK detector. Each layer is made of a monolithic silicon disk with a thin sensitive layer for hit generation. Silicon chip thickness is tuned to match the layer x/X0 to allow a minimal evaluation of material budget effects. Detector layout follows latest Werner's presentation. Alternatively there is a function to build a parametrized detector with equidistant layers (needs recompilation to switch). Detector layout is not yet configurable without recompilation. 

The detector is symmetric at forward and backward directions.


One should get a file `o2sim_HitsFT3.root` by running 
`$ o2-sim -m FT3 -e TGeant3 -g boxgen -n 10 --configKeyValues 'BoxGun.pdg=13 ; BoxGun.eta[0]=-5.0 ; BoxGun.eta[1]=5.0; BoxGun.number=500'`

rpezzi committed:
Adding ALICE EC0 Detector (EndCaps)

Created simple forward layers to get started with ALICE 3 endcaps simulation. Based on the ITSMFT classes, much like the ALICE TRK detector.

rpezzi committed:
Make clang-format happy

mconcas commented on Detectors/Upgrades/ALICE3/EC0/simulation/src/Detector.cxx:
```diff
@@ -0,0 +1,557 @@
+}
+
+//_________________________________________________________________________________________________
+void Detector::Print(std::ostream* os) const
```
I wonder if these methods will be ever used. I am planning to remove them from the ITS3/TRK implementation.
They are quite obsolete. 
Do you think it there is/will be a use case that will make useful to propgate their definition trhougout the years?

mconcas commented on Detectors/Upgrades/ALICE3/EC0/simulation/src/Detector.cxx:
```diff
@@ -0,0 +1,557 @@
+}
+
+//_________________________________________________________________________________________________
+void Detector::Read(std::istream* is)
```
See comment above

mconcas commented on Detectors/Upgrades/ALICE3/EC0/simulation/src/Detector.cxx:
```diff
@@ -0,0 +1,557 @@
+}
+
+//_________________________________________________________________________________________________
+std::ostream& operator<<(std::ostream& os, Detector& p)
```
idem

mconcas commented on Detectors/Upgrades/ALICE3/EC0/simulation/src/Detector.cxx:
```diff
@@ -0,0 +1,557 @@
+}
+
+//_________________________________________________________________________________________________
+std::istream& operator>>(std::istream& is, Detector& r)
```
idem

mconcas reviewed (commented):
Hi @rpezzi, thanks a lot for the PR. To me this looks a positive PR, just one minimal comment about some methods that seem to be not needed (I am removing their counterparts in ITS3 and TRK as well).
I am out of curiosity trying it locally, also to see how things behave in a simulation that includes new beampipe and TRK.
~~As soon as the fullCI is green to me this can be merged~~
As soon as we add also the backwards part of the detector this can be merged.

mconcas commented on Detectors/Upgrades/ALICE3/EC0/simulation/src/Detector.cxx:
```diff
@@ -0,0 +1,557 @@
+  mNumberOfLayers = 10;
+  Float_t sensorThickness = 30.e-4;
+  Float_t layersx2X0 = 1.e-2;
+  std::vector<std::array<Float_t, 5>> layersConfig{
```
Could it be possible to add also the backward part of the detector here?
Configuration should be the symmetric mirroring of the FWD part.

rpezzi replied:
Done

rpezzi committed:
Make codecheck happy

rpezzi committed:
Naming ALICE 3 Forward Tracker (end caps): FT3

rpezzi committed:
Adding symmetric tracker at forward and backward directions

FT3 is now composed of symmetric forward and backward layers.
This commit also removes unused methods.

rpezzi commented:
The detector has been named FT3 and has been mirrored with forward / backward symmetry. Removed unused methods. 
This is how the detector geometry looks like


![image](https://user-images.githubusercontent.com/25225235/109566743-9c661500-7ae4-11eb-8198-bc69e8c1c79f.png) 
![image](https://user-images.githubusercontent.com/25225235/109566858-c0c1f180-7ae4-11eb-84b0-ad90a1e53e47.png)


By the way, pythia events generate hits in all detectors as it should.  My previous test was bogus with o2-sim default fairboxgen.

mconcas commented on Detectors/Upgrades/ALICE3/FT3/simulation/include/FT3Simulation/FT3Layer.h:
```diff
@@ -0,0 +1,68 @@
+  FT3Layer& operator=(const FT3Layer&) = default;
+
+  /// Default destructor
+  ~FT3Layer() = default;
```
```suggestion
  ~FT3Layer() override = default;
```
Should fix fullCI

rpezzi replied:
Yes. I am testing locally... starting to adopt the [Full CI test](https://aliceo2group.github.io/quickstart/devel.html) before pushing: (`ALIBUILD_O2_TESTS=1 aliBuild ...`).

ktf reviewed (dismissed):
Ok for the changes in DataFormats/Headers

mconcas reviewed (approved):
For what concerns the Upgrades part this is fine.

rpezzi commented:
CI is happy! This PR should be good to go.
For reference, `ALIBUILD_O2_TESTS=1 aliBuild ...` does not run checkcode test. Instructions to run checkcode are [here](https://github.com/AliceO2Group/AliceO2/tree/dev/Utilities/Tools). Unfortunately local output was not as readable as the CI. Will keep trusting CI for codecheck.

mconcas commented:
@sawenzel if it is ok with you on the simulation part I think we can merge, thanks.

sawenzel commented on Common/SimConfig/src/SimConfig.cxx:
```diff
@@ -60,7 +60,7 @@ bool SimConfig::resetFromParsedMap(boost::program_options::variables_map const&
     for (int d = DetID::First; d <= DetID::Last; ++d) {
 #ifdef ENABLE_UPGRADES
-      if (d != DetID::IT3 && d != DetID::TRK) {
+      if (d != DetID::IT3 && d != DetID::TRK && d != DetID::FT3) {
```
at some moment it will make sense to talk about a better isolation of Run3 and upgrade detectors. 
Something like predefined list in separate DetIDs ..

mconcas commented:
fullCI seems to be glitched, is there a way to re-trigger the build? @TimoWilken

davidrohr commented:
It will rerun automatically after some time

TimoWilken commented:
@mconcas David is right, but I've marked the build yellow so it re-runs sooner.

rpezzi commented:
Hi @TimoWilken, @mconcas the error currently reported on fullCI is not related to this PR, it is on QC. Last week all CI checks run OK, just before CI broke. I wonder if one could merge this or mark the build yellow again? Thanks.

davidrohr commented:
@Barthelemy : QC is again breaking the FullCI with
```
2021-03-10 03:29:39.372635     QC infologger initialized
2021-03-10 03:29:39.372700     Using config file 'json:///mnt/mesos/sandbox/sandbox/o2-fullci/sw/INSTALLROOT/214c1a6170927dfd0353c030bb72131f1fc93c71/slc8_x86-64/QualityControl/v1.11.0-1/etc/analysisDirect.json'
2021-03-10 03:29:39.372773     Creating a standalone QC topology.
2021-03-10 03:29:39.372803     Generating Data Sampling

 *** Break *** segmentation violation
```
The run is from this morning 3pm. Could you check? I though this was fixed?

rpezzi commented:
@davidrohr @Barthelemy , note that PRs ##5553 and  #5532 ran OK this morning, both on alibuild02, while this PR ran at alientest06.

TimoWilken commented:
@Barthelemy if this issue is fixed in QC v1.12.0, should we use that here? If so, that needs to be changed in alidist. (These checks are still using v1.11.0.)

Barthelemy commented:
It is worked around in QC v1.12 but the alidist PR is itself blocked: https://github.com/alisw/alidist/pull/2887

davidrohr commented:
@TimoWilken @Barthelemy : I am actually wondering, didn't we at some point switch to testing QualityControl/dev instead of the version in alidist?

TimoWilken commented:
@davidrohr, I think it was the other way around -- we test QC against O2@dev. The fullCI checks both currently test against the QC version from alidist (so v1.11.0 right now).

davidrohr commented:
> @davidrohr, I think it was the other way around -- we test QC against O2@dev. The fullCI checks both currently test against the QC version from alidist (so v1.11.0 right now).

Not so sure. Didn't we have the discussion in WP3 for which packets to test the dev branch? And we decided for Readout / etc. to stick to the alibuild version, but for QC we switched to dev. I actually remember seing a PR to change the QC target to dev, but I don't find it any more. Perhaps I am also remembering just wrong...

davidrohr commented:
OK, forget about it, you were right... https://github.com/alisw/ali-bot/pull/931

Barthelemy commented:
I don't mind either ways.

rpezzi commented:
Thanks for the info. How should we proceed with this PR? Notice that it has passed CI tests on March 3. After that all failures on subsequent runs were due to bogus tests.

mconcas commented:
@TimoWilken @davidrohr Hi guys, sorry to bother you, I know you are very busy. Do we have any update on this? Thanks in advance.

TimoWilken commented:
The fullCI error is unrelated. I've put in a PR to fix it -- #5713.

o2-dataflow is also unrelated -- the test in QC has been timing out for lots of PRs. @Barthelemy I assume this is a known problem?
